action/protocol/account/util/util.go

```diff
@@ -30,44 +28,45 @@ func SetNonce(i noncer, state *state.Account) {
 }
 
 // LoadOrCreateAccount either loads an account state or creates an account state
-func LoadOrCreateAccount(sm protocol.StateManager, addr address.Address) (*state.Account, error) {
+func LoadOrCreateAccount(sm protocol.StateManager, addr address.Address, opts ...state.AccountCreationOption) (*state.Account, error) {
 	var (
-		account  state.Account
+		account  = state.NewEmptyAccount()
 		addrHash = hash.BytesToHash160(addr.Bytes())
 	)
-	_, err := sm.State(&account, protocol.LegacyKeyOption(addrHash))
-	if err == nil {
-		return &account, nil
-	}
-	if errors.Cause(err) == state.ErrStateNotExist {
-		if err := account.AddBalance(big.NewInt(0)); err != nil {
-			return nil, err
+	_, err := sm.State(account, protocol.LegacyKeyOption(addrHash))
+	switch errors.Cause(err) {
+	case nil:
+		return account, nil
+	case state.ErrStateNotExist:
+		account, err := state.NewAccount(opts...)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to create state account for %x", addrHash)
 		}
-		account.VotingWeight = big.NewInt(0)
 		if _, err := sm.PutState(account, protocol.LegacyKeyOption(addrHash)); err != nil {
 			return nil, errors.Wrapf(err, "failed to put state for account %x", addrHash)
 		}
-		return &account, nil
+		return account, nil
+	default:
+		return nil, err
 	}
-	return nil, err
 }
 
 // LoadAccount loads an account state by address.Address
-func LoadAccount(sr protocol.StateReader, addr address.Address) (*state.Account, error) {
-	return LoadAccountByHash160(sr, hash.BytesToHash160(addr.Bytes()))
+func LoadAccount(sr protocol.StateReader, addr address.Address, opts ...state.AccountCreationOption) (*state.Account, error) {
+	return LoadAccountByHash160(sr, hash.BytesToHash160(addr.Bytes()), opts...)
 }
 
 // LoadAccountByHash160 loads an account state by 20-byte address
-func LoadAccountByHash160(sr protocol.StateReader, addrHash hash.Hash160) (*state.Account, error) {
-	var account state.Account
-	if _, err := sr.State(&account, protocol.LegacyKeyOption(addrHash)); err != nil {
-		if errors.Cause(err) == state.ErrStateNotExist {
-			account = state.EmptyAccount()
-			return &account, nil
-		}
+func LoadAccountByHash160(sr protocol.StateReader, addrHash hash.Hash160, opts ...state.AccountCreationOption) (*state.Account, error) {
+	account := state.NewEmptyAccount()
+	switch _, err := sr.State(account, protocol.LegacyKeyOption(addrHash)); errors.Cause(err) {
+	case state.ErrStateNotExist:
+		return state.NewAccount(opts...)
+	case nil:
+		return account, nil
```

Comment:
nit: move `case nil` to the first case?
b/c all other `Load()` put nil at first case, nice to be consistent